implement dynamic prompt template and benchmark

zirkelc opened the pull request:
<!--
Thank you for contributing to LangChain.js! Your PR will appear in our next release under the title you set above. Please make sure it highlights your valuable contribution.

To help streamline the review process, please make sure you read our contribution guidelines:
https://github.com/langchain-ai/langchainjs/blob/main/CONTRIBUTING.md

If you are adding an integration (e.g. a new LLM, vector store, or memory), please also read our additional guidelines for integrations:
https://github.com/langchain-ai/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md

Replace this block with a description of the change, the issue it fixes (if applicable), and relevant context.

Finally, we'd love to show appreciation for your contribution - if you'd like us to shout you out on Twitter, please also include your handle below!
-->

This PR is still work in progress!

I have noticed that the performance of the prompts could be significantly improved by leveraging something like Lodash's [`template()`](https://lodash.com/docs/4.17.15#template) function. This function compiles the prompt template string into a javascript source code at instatiation. That means alls placeholders like `{foo}` are actual javascrupt variables which get interpolated by the nodejs engine.

I assembeld a short benchmark comparing a simple prompt with two placeholders and a complex prompt with 100 placeholders:

```sh
zirkelc@Chris-MBP langchainjs % tsx langchain-core/src/prompts/benchmark.ts
Simple prompts:
┌─────────┬─────────────────────────────────┬─────────────┬────────────────────┬──────────┬─────────┐
│ (index) │ Task Name                       │ ops/sec     │ Average Time (ns)  │ Margin   │ Samples │
├─────────┼─────────────────────────────────┼─────────────┼────────────────────┼──────────┼─────────┤
│ 0       │ 'DynamicPromptTemplate: simple' │ '1,640,575' │ 609.5420955809827  │ '±0.93%' │ 1640576 │
│ 1       │ 'PromptTemplate: simple'        │ '761,489'   │ 1313.2150907825624 │ '±4.56%' │ 762757  │
└─────────┴─────────────────────────────────┴─────────────┴────────────────────┴──────────┴─────────┘

Complex prompts:
┌─────────┬──────────────────────────────────┬─────────────┬────────────────────┬──────────┬─────────┐
│ (index) │ Task Name                        │ ops/sec     │ Average Time (ns)  │ Margin   │ Samples │
├─────────┼──────────────────────────────────┼─────────────┼────────────────────┼──────────┼─────────┤
│ 0       │ 'DynamicPromptTemplate: complex' │ '3,748,000' │ 266.80892961438684 │ '±3.67%' │ 3748001 │
│ 1       │ 'PromptTemplate: complex'        │ '1,626,574' │ 614.7888380257064  │ '±1.07%' │ 1626576 │
└─────────┴──────────────────────────────────┴─────────────┴────────────────────┴──────────┴─────────┘
```

As you can see in the results, the `DynamicPromptTemplate` is more than twice as fast as the `PromptTemplate`. 
To run the benchmark, I have used the typescript nodejs runner `tsx` (installed globally) and ran it from the root of the project: `tsx langchain-core/src/prompts/benchmark.ts`

Another advantage of using a template engine is that you have support for nested placeholders (see https://github.com/langchain-ai/langchainjs/discussions/4462):

```ts
type NestedInput = {
    foo: {
        bar: {
            baz: string;
        };
    };
};

const nestedDynamicPrompt = DynamicPromptTemplate.fromTemplate<NestedInput>(`{foo.bar.baz}`);

const formatted = nestedDynamicPrompt.format({
    foo: {
        bar: {
            baz: 'baz',
        },
    },
});
```

Prompts supporting nested variables would remove a lot of unnecessary `RunnableMap` that just exist to unpack an object into the primitive variables which are used as placeholders in a prompt template.

Let me know if you guys are interested in bringing this into LangChain.

jacoblee93 commented:
This looks super interesting!

We want to keep parity with the Python library for prompts for things like serialization, but I'd be down to add this to `experimental` (see e.g. handlebars). Let me know when it's ready for a review!

zirkelc commented:
Hi @jacoblee93 

I renamed the `DynamicPromptTemplate` to `LodashPromptTemplate` and moved it to the experimental folder of the `langchain` package. I also moved the benchmark to the experimental examples and extended the benchmark to also include Handlebars.

Here are the new results including Handlebars:

```sh
Simple prompts:
┌─────────┬────────────────────────────────────┬─────────────┬────────────────────┬──────────┬─────────┐
│ (index) │ Task Name                          │ ops/sec     │ Average Time (ns)  │ Margin   │ Samples │
├─────────┼────────────────────────────────────┼─────────────┼────────────────────┼──────────┼─────────┤
│ 0       │ 'PromptTemplate: simple'           │ '1,298,739' │ 769.9775197500225  │ '±0.86%' │ 1298740 │
│ 1       │ 'LodashPromptTemplate: simple'     │ '1,488,495' │ 671.8194627330353  │ '±0.55%' │ 1488496 │
│ 2       │ 'HandlebarsPromptTemplate: simple' │ '21,873'    │ 45717.535110177705 │ '±1.61%' │ 21874   │
└─────────┴────────────────────────────────────┴─────────────┴────────────────────┴──────────┴─────────┘
Complex prompts:
┌─────────┬─────────────────────────────────────┬──────────┬────────────────────┬──────────┬─────────┐
│ (index) │ Task Name                           │ ops/sec  │ Average Time (ns)  │ Margin   │ Samples │
├─────────┼─────────────────────────────────────┼──────────┼────────────────────┼──────────┼─────────┤
│ 0       │ 'PromptTemplate: complex'           │ '17,932' │ 55763.47660736934  │ '±0.49%' │ 17933   │
│ 1       │ 'LodashPromptTemplate: complex'     │ '40,872' │ 24466.418172924718 │ '±1.45%' │ 40885   │
│ 2       │ 'HandlebarsPromptTemplate: complex' │ '474'    │ 2108994.0252631563 │ '±2.11%' │ 475     │
└─────────┴─────────────────────────────────────┴──────────┴────────────────────┴──────────┴─────────┘
```

In contrast to `HandlebarsPromptTemplate`, the `LodashPromptTemplate` does not implement a custom parser to extract the names of all input variables. This could be easily done, but `lodash.template` support a few different interpolation strings and it has to be decided which style you would like to support. At the moment, I have set the interpolation delimeter to `{foo}` style.

jacoblee93 commented on langchain-core/package.json:
```diff
@@ -56,6 +56,7 @@
     "@langchain/scripts": "~0.0",
     "@swc/core": "^1.3.90",
     "@swc/jest": "^0.2.29",
+    "@types/lodash.template": "^4",
```
Can you remove from core? I think this is a mistake?

jacoblee93 commented on langchain/package.json:
```diff
@@ -1276,6 +1280,7 @@
     "jest": "^29.5.0",
     "jest-environment-node": "^29.6.4",
     "jsdom": "^22.1.0",
+    "lodash.template": "^4.5.0",
```
This should be under `peerDependencies` as well below

jacoblee93 commented on langchain/src/experimental/prompts/lodash.ts:
```diff
@@ -0,0 +1,36 @@
+import template from "lodash.template";
+import { CustomFormatPromptTemplate, CustomFormatPromptTemplateInput } from "./custom_format.js";
+
+export type LodashPromptTemplateInput<RunInput extends InputValues> =
```
Remember to run `yarn format`!

jacoblee93 commented:
A few small comments! Sorry for the delay!

zirkelc commented:
@jacoblee93 no worries! I resolved all comments.
